File: app/src/main/java/com/example/dgtimer/activities/settings/SettingsActivity.java
package com.example.dgtimer.activities.settings;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;

import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;

import com.example.dgtimer.R;
import com.example.dgtimer.databinding.ActivitySettingsBinding;

import static com.example.dgtimer.ApplicationClass.DEFAULT_AMPLITUDE;
import static com.example.dgtimer.ApplicationClass.DEFAULT_VOLUME;
import static com.example.dgtimer.ApplicationClass.GOOGLE_PLAY_LINK;
import static com.example.dgtimer.ApplicationClass.mSharedPreferences;

public class SettingsActivity extends AppCompatActivity {

    private ActivitySettingsBinding binding;
    private SettingsViewModel viewModel;

    private String[] strAlarms;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_settings);

        viewModel = new ViewModelProvider(this).get(SettingsViewModel.class);
        binding.setViewModel(viewModel);
        binding.setLifecycleOwner(this);

        final SharedPreferences.Editor editor = mSharedPreferences.edit();

        viewModel.getVolume().observe(this, new Observer<Integer>() {
            @Override
            public void onChanged(Integer integer) {
                editor.putInt("volume", integer);
                editor.apply();
            }
        });

        viewModel.getAmplitude().observe(this, new Observer<Integer>() {
            @Override
            public void onChanged(Integer integer) {
                editor.putInt("amplitude", integer);
                editor.apply();
            }
        });

        strAlarms = getResources().getStringArray(R.array.strAlarms);
    }

    public void onFinishSettings(View view){
        finish();
    }

    public void onGoReviewsClick(View view){
        Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setData(Uri.parse(GOOGLE_PLAY_LINK));
        startActivity(intent);
    }

    public void onResetBtnClick(View view){
        viewModel.getVolume().setValue(DEFAULT_VOLUME);
        viewModel.getAmplitude().setValue(DEFAULT_AMPLITUDE);
        mSharedPreferences.edit().putInt("alarm", 0).apply();
        viewModel.getCurrentAlarm().set(strAlarms[0]);
    }

    public void onSelectAlarmClick(View view){
        final int[] currentIdx = {0};
        String strCurrentAlarm = viewModel.getCurrentAlarm().get();
        for (int i = 0; i < strAlarms.length; i++) {
            if (strCurrentAlarm.equals(strAlarms[i])){
                currentIdx[0] = i;
                break;
            }
        }

        AlertDialog.Builder builder = new AlertDialog.Builder(SettingsActivity.this);
        builder.setTitle(R.string.selectAlarm).setSingleChoiceItems(strAlarms, currentIdx[0],
                new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                int alarmId = getResources().getIdentifier("alarm_"+which, "raw", getPackageName());
                float fVolume = mSharedPreferences.getInt("volume", DEFAULT_VOLUME) / 100.f;
                viewModel.playAlarm(SettingsActivity.this, alarmId, fVolume);
                currentIdx[0] = which;
            }
        }).setPositiveButton(R.string.save, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                viewModel.getCurrentAlarm().set(strAlarms[currentIdx[0]]);
                mSharedPreferences.edit()
                        .putInt("alarm", currentIdx[0])
                        .apply();
            }
        }).setNegativeButton(R.string.cancel, null).create().show();
    }
}

File: app/src/main/java/com/example/dgtimer/activities/settings/SettingsActivity.kt
package com.example.dgtimer.activities.settings

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.widget.SeekBar
import androidx.activity.viewModels
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.example.dgtimer.AppRater.Companion.launchGooglePlayForRating
import com.example.dgtimer.R
import com.example.dgtimer.databinding.ActivitySettingsBinding
import com.example.dgtimer.utils.AlarmPlayerWrapper
import com.example.dgtimer.utils.Extensions.getPackageInfoCompat
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

@AndroidEntryPoint
class SettingsActivity : AppCompatActivity() {

    private lateinit var binding: ActivitySettingsBinding
    private val viewModel: SettingsViewModel by viewModels()

    private val alarmPlayer by lazy {
        AlarmPlayerWrapper(this)
    }
    private val strAlarms by lazy {
        resources.getStringArray(R.array.strAlarms)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivitySettingsBinding.inflate(layoutInflater)
        setContentView(binding.root)
        initView()
        initObservers()
    }

    override fun onDestroy() {
        alarmPlayer.release()
        super.onDestroy()
    }

    private fun initView() {
        with(binding) {
            sbVolume.setOnSeekBarChangeListener(
                object : SeekBar.OnSeekBarChangeListener {
                    override fun onProgressChanged(
                        seekBar: SeekBar?,
                        value: Int,
                        fromUser: Boolean
                    ) {
                        viewModel.setVolume(value)
                    }

                    override fun onStartTrackingTouch(seekBar: SeekBar?) {}
                    override fun onStopTrackingTouch(seekBar: SeekBar?) {
                        ringAlarm(viewModel.alarm.value)
                    }
                }
            )
            sbVolume.progress = viewModel.volume.value

            sbAmplitude.setOnSeekBarChangeListener(
                object : SeekBar.OnSeekBarChangeListener {
                    override fun onProgressChanged(
                        seekBar: SeekBar?,
                        value: Int,
                        fromUser: Boolean
                    ) {
                        viewModel.setAmplitude(value)
                    }

                    override fun onStartTrackingTouch(seekBar: SeekBar?) {}
                    override fun onStopTrackingTouch(seekBar: SeekBar?) {
                        vibrate()
                    }
                }
            )
            sbAmplitude.progress = viewModel.amplitude.value

            llBtnChooseAlarm.setOnClickListener {
                showChooseAlarmDialog()
            }

            tvAppVersion.text = packageManager
                .getPackageInfoCompat(packageName)
                .versionName

            tvBtnReset.setOnClickListener {
                viewModel.resetAllSettings()
            }

            tvBtnSave.setOnClickListener {
                saveAndFinish()
            }

            tvBtnGoReview.setOnClickListener {
                launchGooglePlayForRating(this@SettingsActivity)
            }
        }
    }

    private fun ringAlarm(alarm: Int) {
        val volume = viewModel.volume.value / 100f
        alarmPlayer.ringAlarm(alarm, volume)
    }

    private fun vibrate() {
        val amplitude = viewModel.amplitude.value
        alarmPlayer.vibrate(amplitude)
    }

    private fun showChooseAlarmDialog() {
        AlertDialog.Builder(this)
            .setTitle(R.string.selectAlarm)
            .setSingleChoiceItems(strAlarms, viewModel.alarm.value) { _, which ->
                viewModel.selectedAlarmIndex = which
                ringAlarm(which)
            }.setPositiveButton(R.string.save) { dialog, _ ->
                viewModel.setAlarm(viewModel.selectedAlarmIndex)
                dialog.dismiss()
            }.setNegativeButton(R.string.cancel, null)
            .show()
    }

    private fun saveAndFinish() {
        viewModel.saveSettings()
        finish()
    }

    private fun initObservers() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                launch {
                    viewModel.volume.collectLatest {
                        binding.tvVolume.text = it.toString()
                        if (binding.sbVolume.progress != it) {
                            binding.sbVolume.progress = it
                        }
                    }
                }

                launch {
                    viewModel.amplitude.collectLatest {
                        binding.tvAmplitude.text = it.toString()
                        if (binding.sbAmplitude.progress != it) {
                            binding.sbAmplitude.progress = it
                        }
                    }
                }

                launch {
                    viewModel.alarm.collect {
                        binding.tvAlarm.text = strAlarms[it]
                    }
                }
            }
        }
    }

    companion object {
        fun createSettingsActivityIntent(
            callerActivity: Activity
        ): Intent = Intent(callerActivity, SettingsActivity::class.java)
    }
}